File: app/src/utils/createLendingPool.ts
import * as anchor from '@coral-xyz/anchor';
import { Connection, PublicKey, SystemProgram } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID, ASSOCIATED_TOKEN_PROGRAM_ID } from '@solana/spl-token';
import fallIdl from '../idl/fall.json';
import { 
  LENDING_SEED,
  LENDING_AUTHORITY_SEED,
  LENDING_TOKEN_SEED,
  BORROW_TOKEN_SEED,
  COLLATERAL_TOKEN_SEED,
} from '../utils/constants';

export async function createLendingPool(
  wallet: any,
  connection: Connection,
  poolPda: PublicKey,
) {
  try {
    console.log('Creating lending pool...');
    const provider = new anchor.AnchorProvider(
      connection,
      wallet,
      { preflightCommitment: "confirmed" }
    );

    const program = new anchor.Program(
      fallIdl,
      provider
    );

    console.log('Fetching pool info...');
    const pool = await program.account.pool.fetch(poolPda);
    console.log('Pool info:', {
      amm: pool.amm.toString(),
      mintA: pool.mintA.toString(),
      mintB: pool.mintB.toString(),
    });

    const mintA = pool.mintA;
    const mintB = pool.mintB;

    // 获取 Lending Pool PDA
    const [lendingPoolPda] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), mintB.toBuffer(), Buffer.from(LENDING_SEED)],
      program.programId
    );

    // 获取 Lending Pool Authority PDA
    const [lendingPoolAuthorityPda] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), mintB.toBuffer(), Buffer.from(LENDING_AUTHORITY_SEED)],
      program.programId
    );

    // 获取 Lending Receipt Token Mint PDA
    const [lendingReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), Buffer.from(LENDING_TOKEN_SEED)],
      program.programId
    );

    // 获取 Borrow Receipt Token Mint PDA
    const [borrowReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintA.toBuffer(), Buffer.from(BORROW_TOKEN_SEED)],
      program.programId
    );

    // 获取 Collateral Receipt Token Mint PDA
    const [collateralReceiptTokenMint] = PublicKey.findProgramAddressSync(
      [poolPda.toBuffer(), mintB.toBuffer(), Buffer.from(COLLATERAL_TOKEN_SEED)],
      program.programId
    );

    // 获取 Lending Pool Token Accounts
    const lendingPoolAccountA = await anchor.utils.token.associatedAddress({
      mint: mintA,
      owner: lendingPoolAuthorityPda
    });

    const lendingPoolAccountB = await anchor.utils.token.associatedAddress({
      mint: mintB,
      owner: lendingPoolAuthorityPda
    });

    console.log('Sending transaction...');
    const createTx = await program.methods.createLendingPool().accounts({
        pool: poolPda,
        mintA: mintA,
        mintB: mintB,
        lendingPool: lendingPoolPda,
        lendingPoolAuthority: lendingPoolAuthorityPda,
        lendingPoolTokenA: lendingPoolAccountA,
        lendingPoolTokenB: lendingPoolAccountB,
        lendingReceiptTokenMint: lendingReceiptTokenMint,
        borrowReceiptTokenMint: borrowReceiptTokenMint,
        collateralReceiptTokenMint: collateralReceiptTokenMint,
        payer: provider.wallet.publicKey,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
        rent: web3.SYSVAR_RENT_PUBKEY,    
      }).rpc();

    // 等待交易确认
    await connection.confirmTransaction(createTx, 'confirmed');

    console.log('Step 2: Initializing lending pool...');
    const initTx = await program.methods.initLendingPool().accounts({
        pool: poolPda,
        mintA: mintA,
        mintB: mintB,
        lendingPool: lendingPoolPda,
        lendingPoolAuthority: lendingPoolAuthorityPda,
        lendingPoolTokenA: lendingPoolAccountA,
        lendingPoolTokenB: lendingPoolAccountB,
        lendingReceiptTokenMint: lendingReceiptTokenMint,
        borrowReceiptTokenMint: borrowReceiptTokenMint,
        collateralReceiptTokenMint: collateralReceiptTokenMint,
        payer: provider.wallet.publicKey,
        tokenProgram: TOKEN_PROGRAM_ID,
        associatedTokenProgram: ASSOCIATED_TOKEN_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
        rent: web3.SYSVAR_RENT_PUBKEY,    
    }).rpc();

    console.log('Init lending pool tx:', initTx);

    const [createTxInfo, initTxInfo] = await Promise.all([
      connection.getTransaction(createTx, {
        commitment: 'confirmed',
        maxSupportedTransactionVersion: 0
      }),
      connection.getTransaction(initTx, {
        commitment: 'confirmed',
        maxSupportedTransactionVersion: 0
      })
    ]);

    // 打印日志
    console.log("Create Transaction Logs:");
    createTxInfo?.meta?.logMessages?.forEach(log => {
      console.log(log);
    });

    console.log("Init Transaction Logs:");
    initTxInfo?.meta?.logMessages?.forEach(log => {
      console.log(log);
    });

    return {
      createTx,
      initTx,
      lendingPoolPda,
      lendingPoolAuthorityPda,
      lendingReceiptTokenMint,
      borrowReceiptTokenMint,
      collateralReceiptTokenMint,
      lendingPoolAccountA,
      lendingPoolAccountB
    };
  } catch (error) {
    console.error('Error in createLendingPool:', error);
    throw error;
  }
}

File: programs/fall/src/instructions/create_lending_pool.rs
use anchor_lang::prelude::*;
use anchor_spl::{
    associated_token::AssociatedToken,
    token::{Mint, Token, TokenAccount},
};

use crate::{
    constants::{ LENDING_SEED,LENDING_AUTHORITY_SEED},
    state::{ Pool,LendingPool},
};

#[derive(Accounts)]
pub struct CreateLendingPool<'info> {
    #[account(
        seeds = [
            pool.amm.as_ref(),
            pool.mint_a.key().as_ref(),
            pool.mint_b.key().as_ref(),
        ],
        bump,
        has_one = mint_a,
        has_one = mint_b,
    )]
    pub pool: Box<Account<'info, Pool>>,

    #[account(
        constraint = mint_a.key() == pool.mint_a
    )]
    pub mint_a: Box<Account<'info, Mint>>,
    #[account(
        constraint = mint_b.key() == pool.mint_b
    )]
    pub mint_b: Box<Account<'info, Mint>>,

    #[account(
        init,
        payer = payer,
        space = LendingPool::LEN,
        seeds = [
            pool.key().as_ref(),
            LENDING_SEED,
        ],
        bump,
    )]
    pub lending_pool: Box<Account<'info, LendingPool>>,

    /// CHECK: Read only authority
    #[account(
        seeds = [
            pool.key().as_ref(),
            LENDING_AUTHORITY_SEED,
        ],
        bump,
    )]
    pub lending_pool_authority: AccountInfo<'info>,

    #[account(
        init,
        payer = payer,
        associated_token::mint = mint_a,
        associated_token::authority = lending_pool_authority,
    )]
    pub lending_pool_token_a: Box<Account<'info, TokenAccount>>,

    #[account(
        init,
        payer = payer,
        associated_token::mint = mint_b,
        associated_token::authority = lending_pool_authority,
    )]
    pub lending_pool_token_b: Box<Account<'info, TokenAccount>>,

    #[account(mut)]
    pub payer: Signer<'info>,

    pub token_program: Program<'info, Token>,
    pub associated_token_program: Program<'info, AssociatedToken>,
    pub system_program: Program<'info, System>,
}

pub fn create_lending_pool(ctx: Context<CreateLendingPool>) -> Result<()> {
    let lending_pool = &mut ctx.accounts.lending_pool;
    lending_pool.pool = ctx.accounts.pool.key();
    lending_pool.min_collateral_ratio = 20000;
    lending_pool.borrow_interest_accumulator_block_height = 0;
    lending_pool.borrow_interest_accumulator = 0;
    lending_pool.share_lending_block_height = 0;
    lending_pool.share_lending_accumulator = 0;
    // lending_pool.mint_a = ctx.accounts.mint_a.key();
    // lending_pool.mint_b = ctx.accounts.mint_b.key();

    Ok(())
}